core/bloom.go

```diff
@@ -0,0 +1,358 @@
+	}
+	info += fmt.Sprintf("error rate: %f, ", b.opts.errorRate)
+	info += fmt.Sprintf("capacity: %d, ", b.opts.capacity)
+	info += fmt.Sprintf("total bits reserved: %d, ", b.opts.bits)
+	info += fmt.Sprintf("bits per element: %f, ", b.opts.bpe)
+	info += fmt.Sprintf("hash functions: %d", len(b.opts.hashFns))
+
+	return info
+}
+
+// add adds a new entry for `value` in the filter. It hashes the given
+// value and sets the bit of the underlying bitset. Returns "-1" in
+// case of errors, "0" if all the bits were already set and "1" if
+// atleast 1 new bit was set.
+func (b *Bloom) add(value string) ([]byte, error) {
+	// We're sure that empty values will be handled upper functions itself.
+	// This is just a property check for the bloom struct.
+	if value == "" {
+		return RESP_MINUS_1, errEmptyValue
+	}
+
+	// Get the indexes where bits are supposed to be set
+	indexes, err := b.opts.getIndexes(value)
+	if err != nil {
+		fmt.Println("error in getting indexes for value:", value, "err:", err)
+		return RESP_MINUS_1, errUnableToHash
+	}
+
+	// Set the bits and keep a count of already set ones
+	count := 0
+	for _, v := range indexes {
+		if isBitSet(b.bitset, int(v)) {
+			count++
+		} else {
+			setBit(b.bitset, int(v))
+		}
+	}
+
+	if count == len(indexes) {
+		// All the bits were already set, return 0 in that case.
+		return RESP_ZERO, nil
+	}
+
+	return RESP_ONE, nil
+}
+
+// exists checks if the given `value` exists in the filter or not.
+// It hashes the given value and checks if the bits are set or not in
+// the underlying bitset. Returns "-1" in case of errors, "0" if the
+// element surely does not exist in the filter, and "1" if the element
+// may or may not exist in the filter.
+func (b *Bloom) exists(value string) ([]byte, error) {
+	// We're sure that empty values will be handled upper functions itself.
+	// This is just a property check for the bloom struct.
+	if value == "" {
+		return RESP_MINUS_1, errEmptyValue
+	}
+
+	// Get the indexes where bits are supposed to be set
+	indexes, err := b.opts.getIndexes(value)
+	if err != nil {
+		fmt.Println("error in getting indexes for value:", value, "err:", err)
+		return RESP_MINUS_1, errUnableToHash
+	}
+
+	// Check if all the bits at given indexes are set or not
+	// Ideally if the element is present, we should find all set bits.
+	for _, v := range indexes {
+		if !isBitSet(b.bitset, int(v)) {
+			// Return with "0" as we found one non-set bit (which is enough to conclude)
+			return RESP_ZERO, nil
+		}
+	}
+
+	// We reached here, which means the element may exist in the filter. Return "1" now.
+	return RESP_ONE, nil
+}
+
+// getIndexes returns a list of indexes of the underlying bit array where
+// bits are supposed to be set (to 1). It uses the set hash function against
+// the given `value` and caps the index with the total number of bits.
+func (opts *BloomOpts) getIndexes(value string) ([]uint64, error) {
+	indexes := make([]uint64, len(opts.hashFns))
```

Comment:
can we not pre-create it and use. Given that the execution is single-threaded, initializing it everytime is just an overhead.

Reply:
Makes sense. We'll need a lock for it in that case, right?